[Quant] onednn backend switch to ideep new api without affacting performance

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90354

**Summary**
Onednn quantization backend switch to new API in `third_party/ideep`.
- `struct forward_params` for conv/deconv are changed. Modify primitive cache accordingly.
- Use new versions of `prepare` and `compute` API. Fp32 and int8 paths separated. The old ones will be deprecated.
- Now `ideep::tensor::reorder_if_differ_in` supports block-to-block reorder. Use it instead of defining a util function `onednn_utils::try_reorder`.
- For new API of transposed convolution, we can use a flag to keep weight desc align with oneDNN thus needless to transpose it explicitly in PyTorch.
- Use `is_channels_last` flag to specify layout of src/dst when querying expected weight desc.

It won't impact correctness. Performance should be unaffected or slightly better.
FBGEMM and QNNPACK backends are not affected.

Performance results are given below.
1. End-to-end performance of static quantized models (from torchvision)
(throughput: fps, higher is better)
![image](https://user-images.githubusercontent.com/12522207/206105879-45c59996-9804-4531-aa1f-dc962e6db5ab.png)

2. Op benchmark of dynamic quantized linear
(Latency: ms, lower is better)
![image](https://user-images.githubusercontent.com/12522207/206124949-77352991-0fda-4285-a484-e20a5797262b.png)

Test method & env:
- Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
- Run multi-instances on a single node. Use one core for each instance.
- Use Jemalloc and Intel OpenMP

**Test plan**
python test/test_quantization.py

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Xia-Weiwen committed:
[Quant] onednn backend switch to ideep new api without affacting performance

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -419,15 +420,14 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> PackedConvWeightsOnednn<
     dims_giohw = with_groups ? ideep::utils::group_dims(dims_iohw, groups) : dims_iohw;
     std::vector<int64_t> perms(dims_giohw.size(), 0); // for permutation of weight
     std::iota(perms.begin(), perms.end(), 0);
-    w_desc = w_desc.transpose(with_groups, with_groups + 1);
```
Explain why this line is removed?

Xia-Weiwen replied:
A flag is added in ideep to control the weight layout to be onednn-style or pytorch-style (One is the transpose of the other): https://github.com/intel/ideep/pull/146
Now we set the flag to have the onednn-style layout, so we do not need to transpose it here anymore.

jgong5 reviewed (commented):
Could you please elaborate on what are the related new changes in the new API that require the corresponding changes in this PR?

Xia-Weiwen commented:
Hi @jgong5 Do we need approvals from other reviewers?

Xia-Weiwen commented:
@pytorchbot rebase

Xia-Weiwen committed:
Update on "[Quant] onednn backend switch to ideep new api without affacting performance"


**Summary**
Onednn quantization backend switch to new API in `third_party/ideep`.
- `struct forward_params` for conv/deconv are changed. Modify primitive cache accordingly.
- Use new versions of `prepare` and `compute` API. Fp32 and int8 paths separated. The old ones will be deprecated.
- Now `ideep::tensor::reorder_if_differ_in` supports block-to-block reorder. Use it instead of defining a util function `onednn_utils::try_reorder`.
- For new API of transposed convolution, we can use a flag to keep weight desc align with oneDNN thus needless to transpose it explicitly in PyTorch.
- Use `is_channels_last` flag to specify layout of src/dst when querying expected weight desc.

It won't impact correctness. Performance should be unaffected or slightly better.
FBGEMM and QNNPACK backends are not affected.

Performance results are given below.
1. End-to-end performance of static quantized models (from torchvision)
(throughput: fps, higher is better)
![image](https://user-images.githubusercontent.com/12522207/206105879-45c59996-9804-4531-aa1f-dc962e6db5ab.png)

2. Op benchmark of dynamic quantized linear
(Latency: ms, lower is better)
![image](https://user-images.githubusercontent.com/12522207/206124949-77352991-0fda-4285-a484-e20a5797262b.png)

Test method & env:
- Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
- Run multi-instances on a single node. Use one core for each instance.
- Use Jemalloc and Intel OpenMP

**Test plan**
python test/test_quantization.py

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 

[ghstack-poisoned]

Xia-Weiwen commented:
@pytorchbot merge

osalpekar commented:
Seeing the following errors: 
```
[2022-12-15T13:57:11.275-08:00] Stderr: fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:569:7: error: no matching function for call to 'prepare'
[CONTEXT] [2022-12-15T13:57:11.275-08:00]       ideep::matmul_forward::prepare</*is_dynamic=*/true>(
[CONTEXT] [2022-12-15T13:57:11.275-08:00]       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2022-12-15T13:57:11.275-08:00] fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:578:5: error: no matching function for call to 'compute'
[CONTEXT] [2022-12-15T13:57:11.275-08:00]     ideep::matmul_forward::compute(params, x, w, b, y, src_scales, src_zero_point);
[CONTEXT] [2022-12-15T13:57:11.275-08:00]     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2022-12-15T13:57:11.275-08:00] fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:580:5: error: no matching function for call to 'compute'
[CONTEXT] [2022-12-15T13:57:11.275-08:00]     ideep::matmul_forward::compute(x, w, b, y,
[CONTEXT] [2022-12-15T13:57:11.275-08:00]     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2022-12-15T13:57:11.275-08:00] Stderr: fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qlinear.cpp:844:7: error: no matching function for call to 'prepare'
[CONTEXT] [2022-12-15T13:57:11.275-08:00]       ideep::matmul_forward::prepare</*is_dynamic=*/false>(
[CONTEXT] [2022-12-15T13:57:11.275-08:00]       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2022-12-15T13:57:11.275-08:00] fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qlinear.cpp:855:5: error: no matching function for call to 'compute'
[CONTEXT] [2022-12-15T13:57:11.275-08:00]     ideep::matmul_forward::compute(x, w, b, y, src_scales, weights_scales,
[CONTEXT] [2022-12-15T13:57:11.275-08:00]     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2022-12-15T13:57:11.276-08:00] Stderr: fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qconv.cpp:1287:7: error: reference to overloaded function could not be resolved; did you mean to call it?
[CONTEXT] [2022-12-15T13:57:11.276-08:00]       ideep::convolution_transpose_forward::compute<false, false>(
[CONTEXT] [2022-12-15T13:57:11.276-08:00]       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2022-12-15T13:57:11.276-08:00] fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qconv.cpp:1288:11: error: expression result unused [-Werror,-Wunused-value]
[CONTEXT] [2022-12-15T13:57:11.276-08:00]           params, src, weights, expected_bias, dst);
[CONTEXT] [2022-12-15T13:57:11.276-08:00]           ^~~~~~
```

More details in D42081737. cc @jerryzh168 @z-a-f

osalpekar commented:
@pytorchbot revert -m "Breaking core pytorch contbuilds internally with function not found errors- more details in D42081737" -c ghfirst

Xia-Weiwen commented:
> @pytorchbot revert -m "Breaking core pytorch contbuilds internally with function not found errors- more details in D42081737" -c ghfirst

Hi @osalpekar I guess the errors were due to version mismatch of `third_party/ideep` in your codebase. Do you have a plan to update `third_party/ideep` to the current version in https://github.com/pytorch/pytorch/tree/master?

Xia-Weiwen commented:
Reopen as https://github.com/pytorch/pytorch/pull/91056
